tensorboard/webapp/widgets/custom_modal/custom_modal.ts

```diff
@@ -0,0 +1,221 @@
+import {TemplatePortal} from '@angular/cdk/portal';
+import {Subscription} from 'rxjs';
+import {isMouseEventInElement} from '../../util/dom';
+
+/**
+ * Enables dynamic creation of modal components.
+ *
+ * # Prerequisites
+ * App root component must define a ViewContainerRef named `vcRef`
+ * e.g.:
+ * ```
+ * // my_app_root.ts
+ * constructor(readonly vcRef: ViewContainerRef) {}
+ * ```
+ *
+ * # Usage
+ * Define a modal using an ng-template:
+ * ```
+ * <ng-template #myModalTemplate>
+ *    <my-awesome-modal-content
+ *      [input1]="input1"
+ *      (output1)="output1"
+ *    >
+ *    </my-awesome-modal-content>
+ * </ng-template>
+ * ```
+ *
+ * Define a ViewChild to reference the template in the component file:
+ * ```
+ * // my_component.ts
+ * ...
+ * @ViewChild('myModalTemplate', {read: TemplateRef})
+ * myModalTemplate!: TemplateRef<unknown>;
+ * ...
+ * ```
+ *
+ * Inject CustomModal into the component
+ * ```
+ * // my_component.ts
+ * ...
+ * constructor(private readonly customModal: CustomModal) {}
+ * ...
+ * ```
+ *
+ * To create a modal, call createNextToElement():
+ * ```
+ * // my_component.ts
+ * ...
+ * onSomeButtonClick(mouseEvent: MouseEventj) {
+ *   this.customModal.createNextToElement(this.myModalTemplate, mouseEvent.target as HTMLElement);
+ * }
+ * ...
+ * ```
+ *
+ * # Testing
+ * A similar setup is required for testing modal behavior from another component.
+ *
+ * Define a separate root component to hold the modal container and ref:
+ * ```
+ * @Component({
+ *   ...,
+ *   template: `<actual-component-to-test></actual-component-to-test>`
+ * })
+ * class TestableComponent {
+ *   constructor(readonly vcRef: ViewContainerRef) {}
+ *   ...
+ * }
+ * ```
+ *
+ * Before each test (e.g. in beforeEach), make sure to add the TestableComponent
+ * to the app component list:
+ * ```
+ * ...
+ * const fixture = TestBed.createComponent(TestableComponent);
+ * const appRef = TestBed.inject(ApplicationRef);
+ * appRef.components.push(fixture.componentRef);
+ * ```
+ */
+
+export class CustomModalRef {
+  overlayRef: OverlayRef;
+  subscriptions: Subscription[];
+
+  constructor(overlayRef: OverlayRef, subscriptions: Subscription[] = []) {
+    this.overlayRef = overlayRef;
+    this.subscriptions = subscriptions;
+  }
+}
+
+@Injectable({providedIn: 'root'})
+export class CustomModal {
+  private customModalRefs: CustomModalRef[] = [];
+
+  constructor(
+    private readonly appRef: ApplicationRef,
+    private readonly overlay: Overlay
+  ) {}
+
+  /** Gets the ViewContainerRef of the app's root component if it exists. */
+  private getModalViewContainerRef(): ViewContainerRef | undefined {
+    const appComponents = this.appRef.components;
+    if (appComponents.length === 0) {
+      // appComponents can potentially be empty in tests.
+      return;
+    }
+
+    const appInstance = appComponents[0].instance;
+    let viewContainerRef: ViewContainerRef = appInstance.vcRef;
```

Comment:
I guess this just happens to work because ngx-color-picker already forced us to do this:
https://github.com/tensorflow/tensorboard/blob/71952d6a3188a553a18af6cba378bba8303b66c6/tensorboard/webapp/app_container.ts#L26

Can you do one of the following?
1) Add documentation to app_container.ts saying that we're also doing this for customModal.
or
2) Personally I think it's reasonable to just ask the other components to pass a ViewContainerRef to CustomModal.createNextToElement() (like how we pass it to TemplatePortal in cc7ebdd933744173d3969748b142c004b354ad86).

Reply:
Actually, since CustomModal is an injectable service, can you just inject ViewContainerRef in its constructor? Or is that something that is only available to Components?